src/antsibull/tagging.py

```diff
@@ -0,0 +1,199 @@
+    """
+    Validate that each collection in tag_data has a repository and a tag
+    associated with it. Return a list of validation errors.
+    """
+    errors = []
+    for name, data in tag_data.items():
+        if not data['repository']:
+            errors.append(
+                f"{name}'s repository is not specified at all in collection-meta.yaml"
+            )
+            continue
+        if not data['tag']:
+            errors.append(
+                f"{name} {data['version']} is not tagged in "
+                f"{data['repository']}"
+            )
+    return errors
+
+
+async def get_collections_tags(
+    data_dir: str, deps_filename: str
+) -> t.Dict[str, t.Dict[str, t.Optional[str]]]:
+    """
+    Iterate over the collections in a CollectionsMetadata file,
+    retrieve their tags, and return a dictionary
+    of collection names mapped to dictionaries containing
+    each collection's 'version' and 'repository' from a DepsFile
+    and the 'tag' that matches.
+    """
+    lib_ctx = app_context.lib_ctx.get()
+
+    deps_filename = os.path.join(data_dir, deps_filename)
+    deps_data = DepsFile(deps_filename).parse()
+    meta_data = CollectionsMetadata(data_dir)
+
+    async with asyncio_pool.AioPool(size=lib_ctx.thread_max) as pool:
+        collection_tags = {}
+        for name, data in meta_data.data.items():
+            collection_tags[name] = pool.spawn_n(
+                _get_collection_tags(deps_data.deps[name], data)
+            )
+        collection_tags = {
+            name: await data for name, data in collection_tags.items()
+        }
+        return collection_tags
+
+
+async def _get_collection_tags(
+    version: str, meta_data=t.Dict[str, t.Optional[str]]
+) -> t.Dict[str, t.Optional[str]]:
+    flog = mlog.fields(func='_get_collection_tags')
+    repository = meta_data.repository
+    data: t.Dict[str, t.Optional[str]] = dict(
+        version=version, repository=repository, tag=None
+    )
+    if meta_data.collection_directory:
+        data['collection_directory'] = meta_data.collection_directory
+    if not repository:
+        flog.debug("'repository' is None. Exitting...")
+        return data
+    tag_version_regex: t.Optional[t.Pattern[str]] = None
+    if meta_data.tag_version_regex:
+        try:
+            tag_version_regex = re.compile(meta_data.tag_version_regex)
+        except re.error as err:
+            flog.fields(err=err).error(f'{tag_version_regex} is an invalid regex')
+            return data
+    async for tag in _get_tags(repository):
+        if _normalize_tag(tag, tag_version_regex) == version:
+            data['tag'] = tag
+            break
+    return data
+
+
+async def _get_tags(repository) -> t.AsyncGenerator[str, None]:
+    flog = mlog.fields(func='_get_tags')
+    args = (
+        'git',
+        'ls-remote',
+        '--refs',
+        '--tags',
+        repository,
+    )
+    flog.debug(f'Running {args}')
+    proc = await asyncio.create_subprocess_exec(
```

Comment:
Why not use `sh` for running commands as all other parts of antsibull do?

Reply:
Yes, but isn't it blocking (in the asyncio sense)?

Reply:
It's blocking (sh does not support asyncio yet, see https://github.com/amoffat/sh/issues/552).